File: modules/crypto.py
import discord
from discord.ext import commands
import requests

class CryptoPriceCog(commands.Cog):
    def __init__(self, bot, api_key):
        self.bot = bot
        self.api_key = api_key

    @commands.command(name='crypto')
    async def crypto_price(self, ctx, symbol='TOP10'):
        """Displays crypto prices. Use a symbol for specific crypto or 'TOP10' for the top 10."""
        if symbol.upper() == 'TOP10':
            message = await self.get_top_10_cryptos()
        else:
            message = await self.get_crypto_price(symbol.upper())

        await ctx.send(message)

    async def get_crypto_price(self, symbol):
        url = "https://pro-api.coinmarketcap.com/v1/cryptocurrency/quotes/latest"
        headers = {'Accepts': 'application/json', 'X-CMC_PRO_API_KEY': self.api_key}
        params = {'symbol': symbol, 'convert': 'PLN'}
        response = requests.get(url, headers=headers, params=params)
        data = response.json()

        if data['status']['error_code'] == 0:
            price = data['data'][symbol]['quote']['PLN']['price']
            return f"The current price of {symbol} is {price:.2f} PLN."
        else:
            return f"Error: Could not fetch the price for {symbol}."

    async def get_top_10_cryptos(self):
        url = "https://pro-api.coinmarketcap.com/v1/cryptocurrency/listings/latest"
        headers = {'Accepts': 'application/json', 'X-CMC_PRO_API_KEY': self.api_key}
        params = {'limit': 10, 'convert': 'PLN'}
        response = requests.get(url, headers=headers, params=params)
        data = response.json()

        if data['status']['error_code'] == 0:
            cryptos = data['data']
            message = "Top 10 Cryptocurrencies:\n"
            for crypto in cryptos:
                name = crypto['name']
                symbol = crypto['symbol']
                price = crypto['quote']['PLN']['price']
                message += f"{name} ({symbol}): {price:.2f} PLN\n"
            return message
        else:
            return "Error: Could not fetch top 10 cryptocurrencies."

async def setup(bot):
    api_key = '8af4164f-ccf2-4463-86f7-aeaf2d6f7f1d'  # Replace with your CoinMarketCap API key
    await bot.add_cog(CryptoPriceCog(bot, api_key))

File: modules/crypto.js
const { SlashCommandBuilder } = require('discord.js');
const { customId, parseCustomId, buttons } = require('../lib/interactions');

const MODULE = 'crypto';
const refreshRow = sym => buttons([{ id: customId(MODULE, 'refresh', sym), label: 'Odśwież', style: 'secondary', emoji: '🔄' }]);

async function fetchJSON(url) {
  const res = await fetch(url, { headers: { 'accept': 'application/json' } });
  if (!res.ok) throw new Error('API error ' + res.status);
  return res.json();
}

async function priceText(sym) {
  if (sym === 'top' || sym === 'top10') {
    const data = await fetchJSON('https://api.coingecko.com/api/v3/coins/markets?vs_currency=usd&order=market_cap_desc&per_page=10&page=1');
    const msg = data.map(c => `${c.name} (${c.symbol.toUpperCase()}): $${c.current_price}`).join('\n');
    return `Top coins:\n${msg}`;
  }
  const data = await fetchJSON(`https://api.coingecko.com/api/v3/simple/price?ids=${encodeURIComponent(sym)}&vs_currencies=usd`);
  const key = Object.keys(data)[0];
  if (!key || !data[key]) throw new Error('Unknown symbol');
  return `${key}: $${data[key].usd}`;
}

module.exports = {
  data: new SlashCommandBuilder()
    .setName('crypto')
    .setDescription('Show cryptocurrency prices (via CoinGecko)')
    .addStringOption(o => o.setName('symbol').setDescription('e.g. bitcoin, ethereum or TOP').setRequired(false)),
  async execute(interaction) {
    const sym = (interaction.options.getString('symbol') || 'top').toLowerCase();
    try {
      const text = await priceText(sym);
      await interaction.client.sendWithLimits(interaction, text, { components: refreshRow(sym) });
    } catch (e) {
      await interaction.reply(`Error: ${e.message || 'failed to fetch'}`);
    }
  },

  // Central component router (main.js) dispatches here for any customId
  // prefixed "crypto:" — see lib/interactions.js and INTERACTIONS.md.
  async handleComponent(interaction) {
    const { payload: sym } = parseCustomId(interaction.customId);
    await interaction.deferUpdate();
    try {
      const text = await priceText(sym);
      await interaction.editReply({ content: text, components: refreshRow(sym) });
    } catch (e) {
      await interaction.editReply({ content: `Error: ${e.message || 'failed to fetch'}`, components: [] });
    }
  }
};
